ldms/src/sampler/zfs_zpool/zfs_zpool.c

```diff
@@ -0,0 +1,624 @@
+/* -*- c-basic-offset: 8 -*- */
+/* Copyright 2022 Lawrence Livermore National Security, LLC
+ * See the top-level COPYING file for details.
+ *
+ * SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause)
+ */
+
+#define _GNU_SOURCE
+/* Next we include the headers to bring in the zfslib in action */
+#include <stdlib.h>
+#include <ctype.h>
+#include <glob.h>
+#include <string.h>
+#include <getopt.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <inttypes.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <libzfs.h>
+#include <libzutil.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include "ldms.h"
+#include "ldmsd.h"
+#include "config.h"
+#include "sampler_base.h"
+#include "zfs_common/zfs_common.h"
+#include <stddef.h>
+
+
+#define SAMP "zpools_stats"
+#define MAX_LINE_LEN 1024
+
+
+#ifndef IFNAMSIZ
+/* from "linux/if.h" */
+#define IFNAMSIZ 16
+#endif
+
+#ifndef PATH_MAX
+#define PATH_MAX 2048
+#endif
+
+#define ZPOOL_VDEV_LIST_SNAME  "zpool_vdev_list"
+
+/* Function prototypes */
+static int get_zpool_stats(zpool_handle_t *zhp, void *data);
+static int get_zpool_count(zpool_handle_t *zhp, void *data);
+static int get_pool_scan_status(nvlist_t *nvroot, const char *pool_name, ldms_mval_t *record_instance);
+
+#define POOL_MEASUREMENT         "zpool_stats"
+#define SCAN_MEASUREMENT         "zpool_scan_stats"
+#define VDEV_MEASUREMENT         "zpool_vdev_stats"
+#define POOL_LATENCY_MEASUREMENT "zpool_latency"
+#define POOL_QUEUE_MEASUREMENT   "zpool_vdev_queue"
+#define MIN_LAT_INDEX   10  /* minimum latency index 10 = 1024ns */
+#define POOL_IO_SIZE_MEASUREMENT        "zpool_io_size"
+#define MIN_SIZE_INDEX  9  /* minimum size index 9 = 512 bytes */
+
+
+typedef int (*stat_printer_f)(nvlist_t *, const char *, const char *);
+
+int complained_about_sync = 0;
+
+static ldmsd_msg_log_f  log_fn;
+static base_data_t      sampler_base;
+static libzfs_handle_t *g_zfs;
+
+static struct {
+        int vdev_list_idx;
+	int vdev_rec_idx;
+} index_store;
+
+static ldms_mval_t list_handle;
+
+typedef enum op_func_type {
+        FUNC_NOFUNCREQ = 0,
+        FUNC_SCRUB     = 1,
+        FUNC_RESILVER  = 2,
+        FUNC_REBUILD   = 3,
+        FUNC_SCAN      = 4
+} op_func_type_;
+
+static const char * const operation_types[] = {
+	[FUNC_NOFUNCREQ] = "none",
+	[FUNC_SCRUB]     = "scrub",
+	[FUNC_RESILVER]  = "resilver",
+	[FUNC_REBUILD]   = "rebuild",
+	[FUNC_SCAN]      = "scan"
+};
+
+/* metric templates for a zpool and scan status if any */
+static struct ldms_metric_template_s zpool_vdev_metrics[] = {
+    {"pool",              0, LDMS_V_CHAR_ARRAY, "", MAX_LINE_LEN },
```

Comment:
Can we choose smaller values for the strings? Maybe check the ZFS code and see what the real limit is for the pool name length. "state" can probably be quite a bit smaller too, etc.

Same thing in the other plugins.

Reply:
looking in ifs...

Reply:
libzfs_impl.h:	char zpool_name[ZFS_MAX_DATASET_NAME_LEN];
sys/fs/zfs.h:#define	ZFS_MAX_DATASET_NAME_LEN 256

Reply:
Set the max at 256 as ZFS and moved the macro in ifs_common.h for all plugins to use.